File: android/src/main/java/com/cloudcare/ft/mobile/sdk/tracker/reactnative/FTRUMModule.java
package com.cloudcare.ft.mobile.sdk.tracker.reactnative;

import com.cloudcare.ft.mobile.sdk.tracker.reactnative.utils.ReactNativeUtils;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.modules.network.OkHttpClientProvider;
import com.facebook.react.modules.network.ReactCookieJarContainer;
import com.ft.sdk.DetectFrequency;
import com.ft.sdk.FTInTakeUrlHandler;
import com.ft.sdk.FTRUMConfig;
import com.ft.sdk.FTRUMGlobalManager;
import com.ft.sdk.FTSdk;
import com.ft.sdk.RUMCacheDiscard;
import com.ft.sdk.garble.bean.AppState;
import com.ft.sdk.garble.bean.NetStatusBean;
import com.ft.sdk.garble.bean.ResourceParams;

import java.util.HashMap;
import java.util.Map;

import kotlin.text.Regex;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;

public class FTRUMModule extends ReactContextBaseJavaModule {

  private static final String DEFAULT_ERROR_TYPE = "reactnative_crash";

  public FTRUMModule(ReactApplicationContext reactContext) {
    super(reactContext);

    OkHttpClientProvider.setOkHttpClientFactory(() -> new OkHttpClient.Builder()
      .addNetworkInterceptor((Interceptor.Chain chain) -> chain.proceed(chain.request()))
      .cookieJar(new ReactCookieJarContainer())
      .build()
    );
  }

  @Override
  public String getName() {
    return "FTReactNativeRUM";
  }

  @ReactMethod
  public void setConfig(ReadableMap context, Promise promise) {
    Map<String, Object> map = context.toHashMap();
    String rumAppId = (String) map.get("androidAppId");
    Double sampleRate = (Double) map.get("sampleRate");
    Boolean enableNativeUserAction = (Boolean) map.get("enableNativeUserAction");
    Boolean enableNativeUserView = (Boolean) map.get("enableNativeUserView");
    Boolean enableNativeUserResource = (Boolean) map.get("enableNativeUserResource");
    Boolean enableResourceHostIP = (Boolean) map.get("enableResourceHostIP");
    Boolean enableTrackNativeCrash = (Boolean) map.get("enableTrackNativeCrash");
    Boolean enableTrackNativeAppANR = (Boolean) map.get("enableTrackNativeAppANR");
    Boolean enableTrackNativeFreeze = (Boolean) map.get("enableTrackNativeFreeze");
    Double nativeFreezeDurationMs = (Double) map.get("nativeFreezeDurationMs");
    Integer monitorType = ReactNativeUtils.convertToNativeInt(map.get("errorMonitorType"));
    Integer deviceMonitorType = ReactNativeUtils.convertToNativeInt(map.get("deviceMonitorType"));
    Integer detectFrequency = ReactNativeUtils.convertToNativeInt(map.get("detectFrequency"));
    Map<String, Object> globalContext = (Map<String, Object>) map.get("globalContext");
    Integer rumCacheLimitCount = ReactNativeUtils.convertToNativeInt(map.get("rumCacheLimitCount"));
    Integer rumDiscardStrategy = ReactNativeUtils.convertToNativeInt(map.get("rumDiscardStrategy"));

    FTRUMConfig rumConfig = new FTRUMConfig().setRumAppId(rumAppId);
    if (sampleRate != null) {
      rumConfig.setSamplingRate(sampleRate.floatValue());
    }
    if (enableNativeUserAction != null) {
      rumConfig.setEnableTraceUserAction(enableNativeUserAction);
    }
    if (enableNativeUserView != null) {
      rumConfig.setEnableTraceUserView(enableNativeUserView);
    }
    if (enableNativeUserResource != null) {
      rumConfig.setEnableTraceUserResource(enableNativeUserResource);
    }
    if (enableResourceHostIP != null) {
      rumConfig.setEnableResourceHostIP(enableResourceHostIP);
    }
    if (enableTrackNativeCrash != null) {
      rumConfig.setEnableTrackAppCrash(enableTrackNativeCrash);
    }
    if (enableTrackNativeFreeze != null) {
      if (nativeFreezeDurationMs != null) {
        rumConfig.setEnableTrackAppUIBlock(enableTrackNativeFreeze, nativeFreezeDurationMs.longValue());
      } else {
        rumConfig.setEnableTrackAppUIBlock(enableTrackNativeFreeze);
      }
    }
    if (enableTrackNativeAppANR != null) {
      rumConfig.setEnableTrackAppANR(enableTrackNativeAppANR);
    }
    if (monitorType != null) {
      rumConfig.setExtraMonitorTypeWithError(monitorType);
    }
    if (deviceMonitorType != null) {
      if (detectFrequency != null) {
        DetectFrequency deviceMetricsDetectFrequency;
        switch (detectFrequency) {
          case 1:
            deviceMetricsDetectFrequency = DetectFrequency.FREQUENT;
            break;
          case 2:
            deviceMetricsDetectFrequency = DetectFrequency.RARE;
            break;
          default:
            deviceMetricsDetectFrequency = DetectFrequency.DEFAULT;
            break;
        }
        rumConfig.setDeviceMetricsMonitorType(deviceMonitorType, deviceMetricsDetectFrequency);
      } else {
        rumConfig.setDeviceMetricsMonitorType(deviceMonitorType);
      }
    }
    if (globalContext != null) {
      for (Map.Entry<String, Object> entry : globalContext.entrySet()) {
        rumConfig.addGlobalContext(entry.getKey(), entry.getValue().toString());
      }
    }
    if (BuildConfig.DEBUG) {
      rumConfig.setResourceUrlHandler(new FTInTakeUrlHandler() {
        @Override
        public boolean isInTakeUrl(String url) {
          return ReactNativeUtils.isReactNativeDevUrl(url);
        }
      });
    }
    if (rumCacheLimitCount != null) {
      rumConfig.setRumCacheLimitCount(rumCacheLimitCount);
    }
    if (rumDiscardStrategy != null) {
      RUMCacheDiscard rumCacheDiscard = RUMCacheDiscard.DISCARD;
      if(rumDiscardStrategy == 1){
        rumCacheDiscard = RUMCacheDiscard.DISCARD_OLDEST;
      }
      rumConfig.setRumCacheDiscardStrategy(rumCacheDiscard);
    }
    FTSdk.initRUMWithConfig(rumConfig);
    //LogUtils.d("configCheck","rumConfig:"+new Gson().toJson(rumConfig));
    promise.resolve(null);
  }

  @ReactMethod
  public void startAction(String actionName, String actionType, ReadableMap map, Promise promise) {
    if (map != null) {
      FTRUMGlobalManager.get().startAction(actionName, actionType, map.toHashMap());
    } else {
      FTRUMGlobalManager.get().startAction(actionName, actionType);
    }
    promise.resolve(null);
  }

  @ReactMethod
  public void onCreateView(String viewName, Double duration, Promise promise) {
    FTRUMGlobalManager.get().onCreateView(viewName, duration.longValue());
    promise.resolve(null);
  }

  @ReactMethod
  public void startView(String viewName, ReadableMap map, Promise promise) {
    if (map != null) {
      FTRUMGlobalManager.get().startView(viewName, map.toHashMap());
    } else {
      FTRUMGlobalManager.get().startView(viewName);
    }
    promise.resolve(null);
  }

  @ReactMethod
  public void stopView(ReadableMap map, Promise promise) {
    if (map != null) {
      FTRUMGlobalManager.get().stopView(map.toHashMap());
    } else {
      FTRUMGlobalManager.get().stopView();
    }
    promise.resolve(null);
  }

  @ReactMethod
  public void addError(String stack, String message, ReadableMap map, Promise promise) {
    if (map != null) {
      FTRUMGlobalManager.get().addError(stack, message, DEFAULT_ERROR_TYPE, AppState.RUN, map.toHashMap());
    } else {
      FTRUMGlobalManager.get().addError(stack, message, DEFAULT_ERROR_TYPE, AppState.RUN);
    }
    promise.resolve(null);
  }

  @ReactMethod
  public void addErrorWithType(String errorType, String stack, String message, ReadableMap map, Promise promise) {
    if (map != null) {
      FTRUMGlobalManager.get().addError(stack, message, errorType, AppState.RUN, map.toHashMap());
    } else {
      FTRUMGlobalManager.get().addError(stack, message, errorType, AppState.RUN);
    }
    promise.resolve(null);
  }

  @ReactMethod
  public void startResource(String key, ReadableMap map, Promise promise) {
    if (map != null) {
      FTRUMGlobalManager.get().startResource(key, map.toHashMap());
    } else {
      FTRUMGlobalManager.get().startResource(key);
    }
    promise.resolve(null);
  }

  @ReactMethod
  public void stopResource(String key, ReadableMap map, Promise promise) {
    if (map != null) {
      FTRUMGlobalManager.get().stopResource(key, map.toHashMap());
    } else {
      FTRUMGlobalManager.get().stopResource(key);
    }
    promise.resolve(null);
  }

  @ReactMethod
  public void addResource(String key, ReadableMap resourceContext, ReadableMap metricsContext, Promise promise) {
    HashMap<String, Object> resourceMap = resourceContext.toHashMap();
    String url = (String) resourceMap.get("url");
    HashMap<String, String> responseHeader = (HashMap<String, String>) resourceMap.get("responseHeader");
    HashMap<String, String> requestHeader = (HashMap<String, String>) resourceMap.get("requestHeader");
    String method = (String) resourceMap.get("httpMethod");
    Integer resourceStatus = ReactNativeUtils.convertToNativeInt(resourceMap.get("resourceStatus"));
    String responseBody = (String) resourceMap.get("responseBody");
//    String responseConnection = (String) resourceMap.get("responseConnection");
//    String responseContentType = (String) resourceMap.get("responseContentType");
//    String responseContentEncoding = (String) resourceMap.get("responseContentEncoding");

//    HashMap<String, Object> metricsMap = metricsContext.toHashMap();
//    Long fetchStartTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("fetchStartTime"));
//    Long tcpStartTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("tcpStartTime"));
//    Long tcpEndTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("tcpEndTime"));
//    Long dnsStartTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("dnsStartTime"));
//    Long dnsEndTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("dnsEndTime"));
//    Long responseStartTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("responseStartTime"));
//    Long responseEndTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("responseEndTime"));
//    Long sslStartTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("sslStartTime"));
//    Long sslEndTime = ReactNativeUtils.convertToNativeLong(metricsMap.get("sslEndTime"));

    ResourceParams params = new ResourceParams();
    params.url =url != null ? url : "";
    params.responseHeader = responseHeader != null ? responseHeader.toString() : "";
    params.resourceMethod = method;
    params.requestHeader = requestHeader != null ? requestHeader.toString() : "";
    params.resourceStatus = resourceStatus != null ? resourceStatus : 0;
    params.responseBody = responseBody != null ? responseBody : "";
//    params.responseConnection = responseConnection != null ? responseConnection : "";
//    params.responseContentType = responseContentType != null ? responseContentType : "";
//    params.responseContentEncoding = responseContentEncoding != null ? responseContentEncoding : "";

    //not implement
    NetStatusBean netStatusBean = new NetStatusBean();
//    netStatusBean.fetchStartTime = (fetchStartTime != null ? fetchStartTime : -1L);
//    netStatusBean.tcpStartTime = (tcpStartTime != null ? tcpStartTime : -1L);
//    netStatusBean.tcpEndTime = (tcpEndTime != null ? tcpEndTime : -1L);
//    netStatusBean.dnsStartTime = (dnsStartTime != null ? dnsStartTime : -1L);
//    netStatusBean.dnsEndTime = (dnsEndTime != null ? dnsEndTime : -1L);
//    netStatusBean.responseStartTime = (responseStartTime != null ? responseStartTime : -1L);
//    netStatusBean.responseEndTime = (responseEndTime != null ? responseEndTime : -1L);
//    netStatusBean.sslStartTime = (sslStartTime != null ? sslStartTime : -1L);
//    netStatusBean.sslEndTime = (sslEndTime != null ? sslEndTime : -1L);

    FTRUMGlobalManager.get().addResource(key, params, netStatusBean);
    promise.resolve(null);
  }

}

File: android/src/main/java/com/cloudcare/ft/mobile/sdk/tracker/reactnative/FTRUMModule.kt
package com.cloudcare.ft.mobile.sdk.tracker.reactnative

import com.cloudcare.ft.mobile.sdk.tracker.reactnative.utils.ReactNativeUtils
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.modules.network.OkHttpClientProvider
import com.facebook.react.modules.network.ReactCookieJarContainer
import com.ft.sdk.DetectFrequency
import com.ft.sdk.FTRUMConfig
import com.ft.sdk.FTRUMGlobalManager
import com.ft.sdk.FTSdk
import com.ft.sdk.garble.bean.AppState
import com.ft.sdk.garble.bean.NetStatusBean
import com.ft.sdk.garble.bean.ResourceParams
import okhttp3.Interceptor
import okhttp3.OkHttpClient

class FTRUMModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {
  companion object {
    val RN_DEV_INNER_URL_REGEX = arrayOf(
      Regex("^http://((10|172|192).[0-9]+.[0-9]+.[0-9]+|localhost|127.0.0.1):808[0-9]/logs$"),//expo
      Regex("^http://localhost:808[0-9]/(hot|symbolicate|message|inspector).*$")//rn
    )
    private const val DEFAULT_ERROR_TYPE = "reactnative_crash"
  }

  init {
    OkHttpClientProvider.setOkHttpClientFactory {
      OkHttpClient.Builder()
        .addNetworkInterceptor(Interceptor { chain ->
          chain.proceed(chain.request())
        })
        .cookieJar(ReactCookieJarContainer())
        .build()
    }

  }

  override fun getName(): String {
    return "FTReactNativeRUM"
  }

  //是否是 debug 发送网络请求
  private fun isDevUrl(text: String, regexArray: Array<Regex>): Boolean {
    return regexArray.any { it.matches(text) }
  }

  @ReactMethod
  fun setConfig(context: ReadableMap, promise: Promise) {
    val map = context.toHashMap()
    val rumAppId = map["androidAppId"] as String
    val sampleRate = map["sampleRate"] as Double?
    val enableNativeUserAction = map["enableNativeUserAction"] as Boolean?
    val enableNativeUserView = map["enableNativeUserView"] as Boolean?
    val enableNativeUserResource = map["enableNativeUserResource"] as Boolean?
    val enableResourceHostIP = map["enableResourceHostIP"] as Boolean?
    val enableTrackNativeCrash = map["enableTrackNativeCrash"] as Boolean?
    val enableTrackNativeAppANR = map["enableTrackNativeAppANR"] as Boolean?
    val enableTrackNativeFreeze = map["enableTrackNativeFreeze"] as Boolean?
    val nativeFreezeDurationMs = map["nativeFreezeDurationMs"] as Double?
    val monitorType = ReactNativeUtils.convertToNativeInt(map["errorMonitorType"])
    val deviceMonitorType = ReactNativeUtils.convertToNativeInt(map["deviceMonitorType"])
    val detectFrequency = ReactNativeUtils.convertToNativeInt(map["detectFrequency"])
    val globalContext = map["globalContext"] as HashMap<String, Any>?

    val rumConfig = FTRUMConfig().setRumAppId(rumAppId)
    if (sampleRate != null) {
      rumConfig.samplingRate = sampleRate.toFloat()
    }

    if (enableNativeUserAction != null) {
      rumConfig.isEnableTraceUserAction = enableNativeUserAction
    }

    if (enableNativeUserView != null) {
      rumConfig.isEnableTraceUserView = enableNativeUserView
    }

    if (enableNativeUserResource != null) {
      rumConfig.isEnableTraceUserResource = enableNativeUserResource
    }
    if (enableResourceHostIP != null) {
      rumConfig.isEnableResourceHostIP = enableResourceHostIP
    }

    if (enableTrackNativeCrash != null) {
      rumConfig.isEnableTrackAppCrash = enableTrackNativeCrash
    }

    if (enableTrackNativeFreeze != null) {
      if (nativeFreezeDurationMs != null) {
        rumConfig.setEnableTrackAppUIBlock(enableTrackNativeFreeze, nativeFreezeDurationMs.toLong())
      } else {
        rumConfig.isEnableTrackAppUIBlock = enableTrackNativeFreeze
      }
    }

    if (enableTrackNativeAppANR != null) {
      rumConfig.isEnableTrackAppANR = enableTrackNativeAppANR
    }

    if (monitorType != null) {
      rumConfig.extraMonitorTypeWithError = monitorType
    }
    if (deviceMonitorType != null) {

      if (detectFrequency != null) {
        val deviceMetricsDetectFrequency: DetectFrequency = when (detectFrequency) {
          0 -> DetectFrequency.DEFAULT
          1 -> DetectFrequency.FREQUENT
          2 -> DetectFrequency.RARE
          else -> DetectFrequency.DEFAULT
        }
        rumConfig.setDeviceMetricsMonitorType(deviceMonitorType, deviceMetricsDetectFrequency)

      } else {
        rumConfig.deviceMetricsMonitorType = deviceMonitorType;
      }
    }

    globalContext?.let {
      for ((key, value) in it) {
        rumConfig.addGlobalContext(key, value.toString())
      }
    }
    if (BuildConfig.DEBUG) {
      rumConfig.setResourceUrlHandler { url ->
        return@setResourceUrlHandler isDevUrl(url, RN_DEV_INNER_URL_REGEX)
      }
    }

    FTSdk.initRUMWithConfig(rumConfig)
    promise.resolve(null)
  }

  @ReactMethod
  fun startAction(actionName: String, actionType: String, map: ReadableMap?, promise: Promise) {
    if (map != null) {
      FTRUMGlobalManager.get().startAction(actionName, actionType, map.toHashMap())
    } else {
      FTRUMGlobalManager.get().startAction(actionName, actionType)
    }
    promise.resolve(null)

  }

  @ReactMethod
  fun onCreateView(viewName: String, duration: Double, promise: Promise) {
    FTRUMGlobalManager.get().onCreateView(viewName, duration.toLong())
    promise.resolve(null)
  }

  @ReactMethod
  fun startView(viewName: String, map: ReadableMap?, promise: Promise) {
    if (map != null) {
      FTRUMGlobalManager.get().startView(viewName, map.toHashMap())
    } else {
      FTRUMGlobalManager.get().startView(viewName)
    }
    promise.resolve(null)
  }

  @ReactMethod
  fun stopView(map: ReadableMap?, promise: Promise) {
    if (map != null) {
      FTRUMGlobalManager.get().stopView(map.toHashMap())
    } else {
      FTRUMGlobalManager.get().stopView()
    }
    promise.resolve(null)
  }

  @ReactMethod
  fun addError(stack: String, message: String, map: ReadableMap?, promise: Promise) {
    if (map != null) {
      FTRUMGlobalManager.get()
        .addError(stack, message, DEFAULT_ERROR_TYPE, AppState.RUN, map.toHashMap())
    } else {
      FTRUMGlobalManager.get().addError(stack, message, DEFAULT_ERROR_TYPE,AppState.RUN)
    }
    promise.resolve(null)
  }

  @ReactMethod
  fun addErrorWithType(
    errorType: String,
    stack: String,
    message: String,
    map: ReadableMap?,
    promise: Promise
  ) {
    if (map != null) {
      FTRUMGlobalManager.get()
        .addError(stack, message, errorType, AppState.RUN, map.toHashMap())
    } else {
      FTRUMGlobalManager.get().addError(stack, message, errorType, AppState.RUN)
    }
    promise.resolve(null)
  }

  @ReactMethod
  fun startResource(key: String, map: ReadableMap?, promise: Promise) {
    if (map != null) {
      FTRUMGlobalManager.get().startResource(key, map.toHashMap())
    } else {
      FTRUMGlobalManager.get().startResource(key)

    }
    promise.resolve(null)
  }

  @ReactMethod
  fun stopResource(key: String, map: ReadableMap?, promise: Promise) {
    if (map != null) {
      FTRUMGlobalManager.get().stopResource(key, map.toHashMap())
    } else {
      FTRUMGlobalManager.get().stopResource(key)
    }
    promise.resolve(null)
  }

  @ReactMethod
  fun addResource(
    key: String,
    resourceContext: ReadableMap,
    metricsContext: ReadableMap,
    promise: Promise
  ) {
    val resourceMap = resourceContext.toHashMap()
    val url = resourceMap["url"] as String?
    val responseHeader = resourceMap["responseHeader"] as HashMap<String, String>?
    val requestHeader = resourceMap["requestHeader"] as HashMap<String, String>?
    val method = resourceMap["resourceMethod"] as String?
    val resourceStatus = ReactNativeUtils.convertToNativeInt(resourceMap["resourceStatus"])
    val responseBody = resourceMap["responseBody"] as String?
    val responseConnection = resourceMap["responseConnection"] as String?
    val responseContentType = resourceMap["responseContentType"] as String?
    val responseContentEncoding = resourceMap["responseContentEncoding"] as String?

    val metricsMap = metricsContext.toHashMap()
    val fetchStartTime = ReactNativeUtils.convertToNativeLong(metricsMap["fetchStartTime"])
    val tcpStartTime = ReactNativeUtils.convertToNativeLong(metricsMap["tcpStartTime"])
    val tcpEndTime = ReactNativeUtils.convertToNativeLong(metricsMap["tcpEndTime"])
    val dnsStartTime = ReactNativeUtils.convertToNativeLong(metricsMap["dnsStartTime"])
    val dnsEndTime = ReactNativeUtils.convertToNativeLong(metricsMap["dnsEndTime"])
    val responseStartTime = ReactNativeUtils.convertToNativeLong(metricsMap["responseStartTime"])
    val responseEndTime = ReactNativeUtils.convertToNativeLong(metricsMap["responseEndTime"])
    val sslStartTime = ReactNativeUtils.convertToNativeLong(metricsMap["sslStartTime"])
    val sslEndTime = ReactNativeUtils.convertToNativeLong(metricsMap["sslEndTime"])

    val params = ResourceParams()
    params.responseHeader = responseHeader.toString()
    params.resourceMethod = method
    params.requestHeader = requestHeader.toString()
    params.resourceStatus = resourceStatus ?: 0
    params.responseBody = responseBody ?: ""
    params.responseConnection = responseConnection ?: ""
    params.responseContentType = responseContentType ?: ""
    params.responseContentEncoding = responseContentEncoding ?: ""
    params.url = url ?: ""

    val netStatusBean = NetStatusBean()
    netStatusBean.fetchStartTime = fetchStartTime ?: -1
    netStatusBean.tcpStartTime = tcpStartTime ?: -1
    netStatusBean.tcpEndTime = tcpEndTime ?: -1
    netStatusBean.dnsStartTime = dnsStartTime ?: -1
    netStatusBean.dnsEndTime = dnsEndTime ?: -1
    netStatusBean.responseStartTime = responseStartTime ?: -1
    netStatusBean.responseEndTime = responseEndTime ?: -1
    netStatusBean.sslStartTime = sslStartTime ?: -1
    netStatusBean.sslEndTime = sslEndTime ?: -1
    FTRUMGlobalManager.get().addResource(key, params, netStatusBean)
    promise.resolve(null)

  }
}
